src/backend/columnar/columnar_customscan.c

```diff
@@ -228,6 +244,177 @@ IsNotIndexPath(Path *path)
+ */
+static Cost
+ColumnarIndexScanAddStartupCost(RelOptInfo *rel, Oid relationId, IndexPath *indexPath)
+{
+	int numberOfColumnsRead = RelationIdGetNumberOfAttributes(relationId);
+
+	/* we would at least read one stripe */
+	return ColumnarPerStripeScanCost(rel, relationId, numberOfColumnsRead);
+}
+
+
+/*
+ * ColumnarIndexScanAddTotalCost returns additional cost estimated for
+ * index scan described by IndexPath for columnar table with relationId.
+ */
+static Cost
+ColumnarIndexScanAddTotalCost(PlannerInfo *root, RelOptInfo *rel,
+							  Oid relationId, IndexPath *indexPath)
+{
+	int numberOfColumnsRead = RelationIdGetNumberOfAttributes(relationId);
+	Cost perStripeCost = ColumnarPerStripeScanCost(rel, relationId, numberOfColumnsRead);
+
+	/*
+	 * We don't need to pass correct loop count to amcostestimate since we
+	 * will only use index correlation & index selectivity, and loop count
+	 * doesn't have any effect on those two.
+	 */
+	double fakeLoopCount = 1;
+	Cost fakeIndexStartupCost;
+	Cost fakeIndexTotalCost;
+	double fakeIndexPages;
+	Selectivity indexSelectivity;
+	double indexCorrelation;
+	amcostestimate_function amcostestimate = indexPath->indexinfo->amcostestimate;
+	amcostestimate(root, indexPath, fakeLoopCount, &fakeIndexStartupCost,
+				   &fakeIndexTotalCost, &indexSelectivity,
+				   &indexCorrelation, &fakeIndexPages);
+
+	Relation relation = RelationIdGetRelation(relationId);
+	uint64 rowCount = ColumnarTableRowCount(relation);
+	RelationClose(relation);
+	double estimatedRows = rowCount * indexSelectivity;
+
+	/*
+	 * In the worst case (i.e no correlation between the column & the index),
+	 * we need to read a different stripe for each row.
+	 */
+	double maxStripeReadCount = estimatedRows;
+
+	/*
+	 * In the best case (i.e the column is fully correlated with the index),
+	 * we wouldn't read the same stripe again and again thanks
+	 * to locality.
+	 */
+	double avgStripeRowCount =
+		rowCount / (double) ColumnarTableStripeCount(relationId);
+	double minStripeReadCount = estimatedRows / avgStripeRowCount;
+
+	/*
+	 * While being close to 0 means low correlation, being close to -1 or +1
+	 * means high correlation. For index scans on columnar tables, it doesn't
+	 * matter if the column and the index are "correlated" (+1) or
+	 * "anti-correlated" (-1) since both help us avoiding from reading the
+	 * same stripe again and again.
+	 */
+	double absIndexCorrelation = Abs(indexCorrelation);
+
+	/*
+	 * To estimate the number of stripes that we need to read, we do linear
+	 * interpolation between minStripeReadCount & maxStripeReadCount. To do
+	 * that, we use complement to 1 of absolute correlation, where being
+	 * close to 0 means high correlation and being close to 1 means low
+	 * correlation.
+	 * In practice, this computation yields a result that makes the index
+	 * quite unfavorable unless index correlation is super high, which is
+	 * mostly not the case in reality.
```

Comment:
Can you word more clearly? What is "mostly not the case in reality"?

There are a lot of cases where we'd expect index correlation to be 1 (or extremely close), like if someone creates an index on the timestamp. They may not really need the index in that case, because chunk group filtering is going to also be very effective. But it is a case where the correlation is 1.

I'd say something like: "In practice, we only want to do an index scan when the correlation is 1 (or extremely close to it), or when the absolute number of tuples returned is very small. Other cases will have a prohibitive cost."

Reply:
done